flake in unit tests of pkg/kubelet/kuberuntime

kubernetes pre-submit unit tests for `pkg/kubelet/kuberuntime` pass some times and fails some times causing the flakiness. Here is one such failure https://k8s-gubernator.appspot.com/build/kubernetes-jenkins/pr-logs/pull/44987/pull-kubernetes-unit/28850/?log#log

Passing:
```

root@s70553-HP-EliteBook-840-G2:~/gopath/src/k8s.io/kubernetes# make test WHAT="./pkg/kubelet/kuberuntime"
Running tests for APIVersion: v1,apps/v1beta1,authentication.k8s.io/v1,authentication.k8s.io/v1beta1,authorization.k8s.io/v1,authorization.k8s.io/v1beta1,autoscaling/v1,autoscaling/v2alpha1,batch/v1,batch/v2alpha1,certificates.k8s.io/v1beta1,extensions/v1beta1,imagepolicy.k8s.io/v1alpha1,policy/v1beta1,rbac.authorization.k8s.io/v1beta1,rbac.authorization.k8s.io/v1alpha1,settings.k8s.io/v1alpha1,storage.k8s.io/v1beta1,storage.k8s.io/v1,federation/v1beta1
+++ [0503 17:58:43] Running tests without code coverage
ok  	k8s.io/kubernetes/pkg/kubelet/kuberuntime	0.124s
```
Failing:
```
root@s70553-HP-EliteBook-840-G2:~/gopath/src/k8s.io/kubernetes# make test WHAT="./pkg/kubelet/kuberuntime"
Running tests for APIVersion: v1,apps/v1beta1,authentication.k8s.io/v1,authentication.k8s.io/v1beta1,authorization.k8s.io/v1,authorization.k8s.io/v1beta1,autoscaling/v1,autoscaling/v2alpha1,batch/v1,batch/v2alpha1,certificates.k8s.io/v1beta1,extensions/v1beta1,imagepolicy.k8s.io/v1alpha1,policy/v1beta1,rbac.authorization.k8s.io/v1beta1,rbac.authorization.k8s.io/v1alpha1,settings.k8s.io/v1alpha1,storage.k8s.io/v1beta1,storage.k8s.io/v1,federation/v1beta1
+++ [0503 17:56:10] Running tests without code coverage
--- FAIL: TestSandboxGC (0.00s)
	kuberuntime_gc_test.go:125: TestCase #0: {description:notready sandboxes without containers for deleted pods should be garbage collected. sandboxes:[{pod:0xc420023800 attempt:0 createdAt:0 state:1}] containers:[] minAge:0 remain:[]}
	kuberuntime_gc_test.go:125: TestCase #1: {description:ready sandboxes without containers for deleted pods should not be garbage collected. sandboxes:[{pod:0xc420023800 attempt:0 createdAt:0 state:0}] containers:[] minAge:0 remain:[0]}
	kuberuntime_gc_test.go:125: TestCase #2: {description:sandboxes for existing pods should not be garbage collected. sandboxes:[{pod:0xc420022f00 attempt:0 createdAt:0 state:0} {pod:0xc420023200 attempt:0 createdAt:0 state:1}] containers:[] minAge:0 remain:[0 1]}
	kuberuntime_gc_test.go:125: TestCase #3: {description:sandbox with containers should not be garbage collected. sandboxes:[{pod:0xc420022f00 attempt:0 createdAt:0 state:1}] containers:[{pod:0xc420022f00 container:0xc4201ec780 sandboxAttempt:0 attempt:0 createdAt:0 state:2}] minAge:0 remain:[0]}
	kuberuntime_gc_test.go:125: TestCase #4: {description:multiple sandboxes should be handled properly. sandboxes:[{pod:0xc420022f00 attempt:1 createdAt:0 state:0} {pod:0xc420022f00 attempt:0 createdAt:0 state:1} {pod:0xc420023200 attempt:1 createdAt:0 state:1} {pod:0xc420023200 attempt:0 createdAt:0 state:1} {pod:0xc420023800 attempt:0 createdAt:0 state:1}] containers:[{pod:0xc420023200 container:0xc4201d17c0 sandboxAttempt:1 attempt:0 createdAt:0 state:2}] minAge:0 remain:[0 2]}
        Error Trace:    kuberuntime_gc_test.go:135
	Error:		"[&PodSandbox{Id:foo2_new_5678_1,Metadata:&PodSandboxMetadata{Name:foo2,Uid:5678,Namespace:new,Attempt:1,},State:SANDBOX_NOTREADY,CreatedAt:0,Labels:map[string]string{io.kubernetes.pod.name: foo2,io.kubernetes.pod.namespace: new,io.kubernetes.pod.uid: 5678,},Annotations:map[string]string{},} &PodSandbox{Id:foo2_new_5678_0,Metadata:&PodSandboxMetadata{Name:foo2,Uid:5678,Namespace:new,Attempt:0,},State:SANDBOX_NOTREADY,CreatedAt:0,Labels:map[string]string{io.kubernetes.pod.name: foo2,io.kubernetes.pod.namespace: new,io.kubernetes.pod.uid: 5678,},Annotations:map[string]string{},} &PodSandbox{Id:foo1_new_1234_1,Metadata:&PodSandboxMetadata{Name:foo1,Uid:1234,Namespace:new,Attempt:1,},State:SANDBOX_READY,CreatedAt:0,Labels:map[string]string{io.kubernetes.pod.name: foo1,io.kubernetes.pod.namespace: new,io.kubernetes.pod.uid: 1234,},Annotations:map[string]string{},}]" should have 2 item(s), but has 3
		
E0503 17:56:19.635869    6170 kuberuntime_manager.go:554] Couldn't make a ref to pod "foo_new(12345678)": 'selfLink was empty, can't make reference'
E0503 17:56:19.636062    6170 image_manager.go:90] Couldn't make a ref to pod foo, container foo1: 'selfLink was empty, can't make reference'
E0503 17:56:19.636236    6170 kuberuntime_container.go:64] Can't make a ref to pod "foo_new(12345678)", container foo1: selfLink was empty, can't make reference
E0503 17:56:19.636292    6170 image_manager.go:90] Couldn't make a ref to pod foo, container foo2: 'selfLink was empty, can't make reference'
E0503 17:56:19.636380    6170 kuberuntime_container.go:64] Can't make a ref to pod "foo_new(12345678)", container foo2: selfLink was empty, can't make reference
E0503 17:56:19.639891    6170 kuberuntime_manager.go:554] Couldn't make a ref to pod "foo_new(12345678)": 'selfLink was empty, can't make reference'
E0503 17:56:19.639929    6170 image_manager.go:90] Couldn't make a ref to pod foo, container init1: 'selfLink was empty, can't make reference'
E0503 17:56:19.640053    6170 kuberuntime_container.go:64] Can't make a ref to pod "foo_new(12345678)", container init1: selfLink was empty, can't make reference
E0503 17:56:19.640194    6170 image_manager.go:90] Couldn't make a ref to pod foo, container foo1: 'selfLink was empty, can't make reference'
E0503 17:56:19.640286    6170 kuberuntime_container.go:64] Can't make a ref to pod "foo_new(12345678)", container foo1: selfLink was empty, can't make reference
E0503 17:56:19.640317    6170 image_manager.go:90] Couldn't make a ref to pod foo, container foo2: 'selfLink was empty, can't make reference'
E0503 17:56:19.640415    6170 kuberuntime_container.go:64] Can't make a ref to pod "foo_new(12345678)", container foo2: selfLink was empty, can't make reference
FAIL
FAIL	k8s.io/kubernetes/pkg/kubelet/kuberuntime	0.121s
Makefile:164: recipe for target 'test' failed
make: *** [test] Error 1
```

cc @kubernetes/sig-node-bugs

--- a/pkg/kubelet/kuberuntime/kuberuntime_gc_test.go
+++ b/pkg/kubelet/kuberuntime/kuberuntime_gc_test.go
@@ -35,15 +35,16 @@ func TestSandboxGC(t *testing.T) {
 	assert.NoError(t, err)
 
 	fakePodGetter := m.containerGC.podGetter.(*fakePodGetter)
-	makeGCSandbox := func(pod *v1.Pod, attempt uint32, state runtimeapi.PodSandboxState, withPodGetter bool) sandboxTemplate {
+	makeGCSandbox := func(pod *v1.Pod, attempt uint32, state runtimeapi.PodSandboxState, withPodGetter bool, createdAt int64) sandboxTemplate {
 		if withPodGetter {
 			// initialize the pod getter
 			fakePodGetter.pods[pod.UID] = pod
 		}
 		return sandboxTemplate{
-			pod:     pod,
-			state:   state,
-			attempt: attempt,
+			pod:       pod,
+			state:     state,
+			attempt:   attempt,
+			createdAt: createdAt,
 		}
 	}
 
@@ -70,7 +71,7 @@ func TestSandboxGC(t *testing.T) {
 		{
 			description: "notready sandboxes without containers for deleted pods should be garbage collected.",
 			sandboxes: []sandboxTemplate{
-				makeGCSandbox(pods[2], 0, runtimeapi.PodSandboxState_SANDBOX_NOTREADY, false),
+				makeGCSandbox(pods[2], 0, runtimeapi.PodSandboxState_SANDBOX_NOTREADY, false, 0),
 			},
 			containers: []containerTemplate{},
 			remain:     []int{},
@@ -78,7 +79,7 @@ func TestSandboxGC(t *testing.T) {
 		{
 			description: "ready sandboxes without containers for deleted pods should not be garbage collected.",
 			sandboxes: []sandboxTemplate{
-				makeGCSandbox(pods[2], 0, runtimeapi.PodSandboxState_SANDBOX_READY, false),
+				makeGCSandbox(pods[2], 0, runtimeapi.PodSandboxState_SANDBOX_READY, false, 0),
 			},
 			containers: []containerTemplate{},
 			remain:     []int{0},
@@ -86,8 +87,8 @@ func TestSandboxGC(t *testing.T) {
 		{
 			description: "sandboxes for existing pods should not be garbage collected.",
 			sandboxes: []sandboxTemplate{
-				makeGCSandbox(pods[0], 0, runtimeapi.PodSandboxState_SANDBOX_READY, true),
-				makeGCSandbox(pods[1], 0, runtimeapi.PodSandboxState_SANDBOX_NOTREADY, true),
+				makeGCSandbox(pods[0], 0, runtimeapi.PodSandboxState_SANDBOX_READY, true, 0),
+				makeGCSandbox(pods[1], 0, runtimeapi.PodSandboxState_SANDBOX_NOTREADY, true, 0),
 			},
 			containers: []containerTemplate{},
 			remain:     []int{0, 1},
@@ -95,7 +96,7 @@ func TestSandboxGC(t *testing.T) {
 		{
 			description: "sandbox with containers should not be garbage collected.",
 			sandboxes: []sandboxTemplate{
-				makeGCSandbox(pods[0], 0, runtimeapi.PodSandboxState_SANDBOX_NOTREADY, false),
+				makeGCSandbox(pods[0], 0, runtimeapi.PodSandboxState_SANDBOX_NOTREADY, false, 0),
 			},
 			containers: []containerTemplate{
 				{pod: pods[0], container: &pods[0].Spec.Containers[0], state: runtimeapi.ContainerState_CONTAINER_EXITED},
@@ -106,15 +107,15 @@ func TestSandboxGC(t *testing.T) {
 			description: "multiple sandboxes should be handled properly.",
 			sandboxes: []sandboxTemplate{
 				// running sandbox.
-				makeGCSandbox(pods[0], 1, runtimeapi.PodSandboxState_SANDBOX_READY, true),
+				makeGCSandbox(pods[0], 1, runtimeapi.PodSandboxState_SANDBOX_READY, true, 1),
 				// exited sandbox without containers.
-				makeGCSandbox(pods[0], 0, runtimeapi.PodSandboxState_SANDBOX_NOTREADY, false),
+				makeGCSandbox(pods[0], 0, runtimeapi.PodSandboxState_SANDBOX_NOTREADY, true, 0),
 				// exited sandbox with containers.
-				makeGCSandbox(pods[1], 1, runtimeapi.PodSandboxState_SANDBOX_NOTREADY, true),
+				makeGCSandbox(pods[1], 1, runtimeapi.PodSandboxState_SANDBOX_NOTREADY, true, 1),
 				// exited sandbox without containers.
-				makeGCSandbox(pods[1], 0, runtimeapi.PodSandboxState_SANDBOX_NOTREADY, false),
+				makeGCSandbox(pods[1], 0, runtimeapi.PodSandboxState_SANDBOX_NOTREADY, true, 0),
 				// exited sandbox without containers for deleted pods.
-				makeGCSandbox(pods[2], 0, runtimeapi.PodSandboxState_SANDBOX_NOTREADY, false),
+				makeGCSandbox(pods[2], 0, runtimeapi.PodSandboxState_SANDBOX_NOTREADY, false, 0),
 			},
 			containers: []containerTemplate{
 				{pod: pods[1], container: &pods[1].Spec.Containers[0], sandboxAttempt: 1, state: runtimeapi.ContainerState_CONTAINER_EXITED},

